servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -78,58 +88,76 @@ private Single<C> p2c(int size, List<Host<ResolvedAddress, C>> hosts, Random ran
                           boolean forceNewConnectionAndReserve, @Nullable ContextMap contextMap) {
         // If there are only two hosts we only try once since there is no chance we'll select different hosts
         // on further iterations.
+        Host<ResolvedAddress, C> failOpenHost = null;
         for (int j = hosts.size() == 2 ? 1 : maxEffort; j > 0; j--) {
             // Pick two random indexes that don't collide. Limit the range on the second index to 1 less than
             // the max value so that if there is a collision we can safety increment. We also increment if
-            // i2 > i1 to avoid biased toward lower numbers since we limited the range by 1.
+            // i2 > i1 to avoid bias toward lower numbers since we limited the range by 1.
             final int i1 = random.nextInt(size);
             int i2 = random.nextInt(size - 1);
             if (i2 >= i1) {
                 ++i2;
             }
+
             Host<ResolvedAddress, C> t1 = hosts.get(i1);
             Host<ResolvedAddress, C> t2 = hosts.get(i2);
-            // Make t1 the preferred host by score to make the logic below a bit cleaner.
-            if (t1.score() < t2.score()) {
-                Host<ResolvedAddress, C> tmp = t1;
-                t1 = t2;
-                t2 = tmp;
-            }
-
-            // Attempt to get a connection from t1 first since it's 'better'. If we can't, then try t2.
-            Single<C> result = selectFromHost(t1, selector, forceNewConnectionAndReserve, contextMap);
-            if (result != null) {
-                return result;
+            final Host.Status t1Status = t1.status(forceNewConnectionAndReserve);
+            final Host.Status t2Status = t2.status(forceNewConnectionAndReserve);
+            // Priority of selection: health > score > failOpen
+            // Only if both hosts are healthy do we consider score.
+            if (t1Status.healthy && t2Status.healthy) {
+                // both are healthy. Select based on score, using t1 if equal.
+                if (t1.score() < t2.score()) {
```

Comment:
Current implementation, doesn't require us to call `scote()` on every selection attempt; decay happens any time we call it, relevant to the time passed. I wonder whether this will always hold true, or we should invoke `score()` on other flows proactively.

Reply:
A `score()` method _shouldn't_ (technically FP math errors can affect this) have external side effects. The score of least-loaded is pretty obvious and in the case of an EWMA, the decay operation is such that calling `.score()` at time `T` yield the same value regardless of how many times it was called before that time (presuming another sample hasn't been added in the interim).